### DISCUSSION TITLE
Constraint Optimization with SingleTaskVariationalGP of binary classification task

### DISCUSSION THREAD

**Discussion by ToennisStef** [meta-pytorch/botorch#2424] [Q&A]
Hi everyone, 
I have a general question about Bayesian Optimization with unknown constraints. 
I want to implement a BO loop with feasibility constraints. I have the discrete labels (0=feasible, 1=infeasible). 
To model the constraint I am using the SingleTaskVariationalGP with a BernoulliLikelihood. 
To implement the constraint into the BO loop I am using the LogConstrainedExpectedImprovement.

!!! Warning no reproducable code!!!
This is generally the implementation:

```
train_x_f = # some fesible input data
train_y_f = # some fesible objective data

train_x_con = # feasible & infesible data
train_con = # tensor constaining the feasibility labels

model_obj = SingleTaskGP(train_X=train_x_f, 
                       train_Y=train_y_f, 
                       train_Yvar=torch.full_like(train_y_f, 0),
                       input_transform= Normalize(train_x_f.shape[-1]),
                       outcome_transform=Standardize(train_y_f.shape[-1])
                       )
    
model_con = SingleTaskVariationalGP(train_X=train_x_con, 
                       train_Y=train_con, 
                       likelihood=gpytorch.likelihoods.BernoulliLikelihood(),
                       input_transform= Normalize(train_x_con.shape[-1]),
                       outcome_transform=Standardize(train_con.shape[-1])
                       )

mll_obj = ExactMarginalLogLikelihood(model_obj.likelihood, model_obj)
mll_con = VariationalELBO(model_con.likelihood, model_con.model, train_y_f.shape[0])

fit_gpytorch_mll(mll_obj)
fit_gpytorch_mll(mll_con)

model = ModelListGP(model_obj, model_con)

constraints = {1: (None, 0.0)}
acq_func = LogConstrainedExpectedImprovement(model, best_f=train_y_f.max(), objective_index = 0, constraints=constraints)

bounds = # Some bounds
NUM_RESTARTS = 10
RAW_SAMPLES = 521

candidates, acq = optimize_acqf(
        acq_function=acq_func,
        bounds=bounds,
        q=BATCH_SIZE,
        num_restarts=NUM_RESTARTS,
        raw_samples=RAW_SAMPLES,
    )

```

For a real BO loop implementation there is obviously a lot missing. But my question is regarding the implementation of the constraint. 
I checked the SingleTaskVariationalGP model and the logits at the position where the training inputs are infeasible are negative. 
But is it in general sufficient to implement the constraint like this? 
Please correct me if i am wrong but to my knowledge, the output of the model_con is a mutlivariate gaussian distribution representing the logit of my constraint and the LogConstraintExpectedImprovment scales the acquisition with the probability of the logit distribution being between minus infinity and zero. Is that correct? 
If my understanding of what is happening is correct the implementation of the binary feasibility constraint shoud be sufficient, correct?
I am happy for any help.

I tried to adjust the implementation from #640  but this issue is about multi-output optimization. There he also used a Variational GP model but implemented a new objective function and scaled it with the "real" probability of feasibility (not the logit distribution) calculated from MC samples and then proceeded with an MC-acquisition function.
But to my understanding if we are already using a VariationalGP model we should be able to use an analytic acquisition function.

**Comment by esantorella** [comment]
This is a really interesting question. There have been some discussions of "failure-aware BO", and that feels similar to me, as does the idea of data missing not-at-random.

Your approach seems like it could work to me. I haven't worked with `LogConstrainedExpectedImprovement` myself, but looking at the docstring, I would think you'd want the constraint to be `{1: (0.5, 1)}`, since feasible points should have predictions closer to 1 than to 0. I would expect all the predictions to be between 0 and 1 since the labels are 0 and 1.

If you have a reproducible example, I could take a look -- I think this would be an interesting example for the BoTorch documentation.
